keps/sig-cloud-provider/20190729-building-without-in-tree-providers.md

```diff
@@ -0,0 +1,232 @@
+  - Enable testing out of tree providers with a simulation of the future removal of the in-tree code.
+  - Enable experimentation with cloud-provider-less clusters
+
+### Non-Goals
+
+- Building the out of tree providers
+- Changing the official Kubernetes release builds
+- Building the e2e tests 
+  - Decoupling cloud providers is a larger problem there and not necessary to test out-of-tree providers or build smaller binaries
+- Mechanisms for migrating to out of tree providers
+  - CSI Migration for in-tree Volumes is already underway in SIG Storage
+  - External Credential Providers is being written / solved in another KEP ([#541](https://github.com/kubernetes/enhancements/issues/541))
+  - CCMs and overall scope for moving out of tree is in [removing-in-tree-providers](keps/sig-cloud-provider/20190125-removing-in-tree-providers.md)
+
+## Proposal
+
+We will add a [build constraints](https://golang.org/pkg/go/build/#hdr-Build_Constraints)
+to the cloud provider code for a pseudo "build tag" specifying not to include
+any in-tree provider code. This will allow compiling the binaries as normal today
+and simulating the removal of this code by specifying the tag at build time and
+triggering the build constraints on the files in these packages.
+
+Some small adjustments may be necessary to the code base to ensure that the
+other packages can build without depending on these packages.
+
+A prototype is available in [kubernetes/kubernetes#80353](https://github.com/kubernetes/kubernetes/pull/80353).
+
+To ensure that this continues to function we will add CI building in this mode,
+and CI running end to end tests against it (see the test plan).
+
+### User Stories [optional]
+
+#### Story 1
+
+As an out of tree cloud provider implementer, I want to develop and test against
+Kubernetes without the in tree providers.
+
+Kubernetes out of tree cloud provider developers will be able to build Kubernetes
+in this mode and build & test their cloud-controller-manager implementations and
+associated tooling against this build in preparation for the actual hard removal
+of the in-tree providers.
+
+#### Story 2
+
+As a [kind](https://github.com/kubernetes-sigs/kind) developer / user I want to 
+use Kubernetes binaries without cloud providers for local clusters.
+
+Developers and users will be able to build local clusters leveraging this mode
+to not pay for cloud providers they are unable to use.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This is implemented using a synthetic `nolegacyproviders` tag in go build
+constraints on the relevant sources. If `GOFLAGS=-tags=nolegacyproviders` is set
+then the legacy cloud provider pacakges will be excluded from the build.
+
+In order to make this work the following additional changes are made:
+
+- Packages that we fully exclude (the legacy provider packages) _must_ contain
+a `doc.go` file or any other file that does NOT contain any code or build 
```

Comment:
Worth specifying standard verbiage for the doc.go, at least if its in K/K? Eg. a comment to the effect that this package is currently only used by cloud provider specific code. It will be removed from K/K as part of the cloud provider extraction effort.

Do we have such packages which are not part of K/K? If so do we have mechanism for putting a doc.go in them?

Reply:
> Worth specifying standard verbiage for the doc.go, at least if its in K/K? Eg. a comment to the effect that this package is currently only used by cloud provider specific code. It will be removed from K/K as part of the cloud provider extraction effort.

SGTM

> Do we have such packages which are not part of K/K? If so do we have mechanism for putting a doc.go in them?

 That is a really good question, currently we do not if we're only considering things that import the in-tree legacy provider packages. Ideally we should not import packages that import these in the future...

Reply:
Just a note that I don't see any `doc.go` verbiage specified here.  I'm comfortable seeing this KEP merge and having the verbiage sorted out over in the prototype PR

eg: https://github.com/kubernetes/kubernetes/pull/80353/files#diff-fe522781fb7f1bf08a17a523c264ffb6 no verbiage at the moment

Reply:
Yeah, I'm not sure what verbiage we want, but it also seems very into the weeds on details that might best be reviewed in the implementation.

Currently the implementation PR merely adds stub doc.go files containing only the license boilerplate and the `package foo` line IFF the package did not already contain a `doc.go`. 

Many of our packages contain them with anything from this sort of stub to actual godoc comments.